### PR TITLE
A utility WorkflowAlgorithmRunner to manage data reduction workflows

### PR DESCRIPTION
**Pull request by soininen**
The purpose of this algorithm is to do dependency resolution, data flow control, and execution management for another algorithm. It is mainly intended to be used later with certain data reduction workflow algorithms at ILL; however, it can run any type of algorithm if needed.

Implements #17480 . For further documentation, see [this](https://github.com/mantidproject/documents/blob/master/Project-Management/ILL/WorkspaceAlgorithmRunner.md).

**To test:**
The algorithm needs a bit of setup. I'll use the Python example from the documentation here; it should provide a nice starting point for testing.

``` python
# This is our initial workspace. We want to scale it first by 100
# and then by 0.42
CreateSingleValuedWorkspace(OutputWorkspace='ws2', DataValue=1.0)

# Setup the runs for the Scale algorithm
setupTable = WorkspaceFactoryImpl.Instance().createTable()
setupTable.addColumn('str', 'Run name')
# The rest of the columns can be in arbitrary order
setupTable.addColumn('str', 'InputWorkspace')
setupTable.addColumn('double', 'Factor')
setupTable.addColumn('str', 'OutputWorkspace')
row = {
    'Run name': '1st run',
    'InputWorkspace': '2nd run', # Takes the input of '2nd run'.
    'Factor': 0.42,
    'OutputWorkspace': '"ws1"' # Forced output. Otherwise Scale fails to execute.
}
setupTable.addRow(row)
row = {
    'Run name': '2nd run',
    'InputWorkspace': "'ws2'", # Hard-coded input
    'Factor': 100,
    'OutputWorkspace': 'scaled_ws2'
}
setupTable.addRow(row)
AnalysisDataServiceImpl.Instance().addOrReplace('setupTable', setupTable)

# Map the Scale algorithm's OutputWorkspace to InputWorkspace
ioMap = WorkspaceFactoryImpl.Instance().createTable()
ioMap.addColumn('str', 'InputWorkspace')
ioMap.addRow({'InputWorkspace': 'OutputWorkspace'})
AnalysisDataServiceImpl.Instance().addOrReplace('ioMapTable', ioMap)

# Execute the algorithm
WorkflowAlgorithmRunner('Scale', SetupTable=setupTable, InputOutputMap=ioMap)

# Print some results
print('Original input value: {0}'.format(mtd['ws2'].dataY(0)[0]))
print('After scaling by 100: {0}'.format(mtd['scaled_ws2'].dataY(0)[0]))
print('After further scaling by 0.42: {0}'.format(mtd['ws1'].dataY(0)[0]))
```

The above script should print results 1, 100, and 42.

_Release notes were updated._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/WorkflowAlgorithmRunner.cpp [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+  std::deque<size_t> queue;
+  for (size_t i = 0; i < setupTable->rowCount(); ++i) {
+    configureRow(setupTable, propertyTable, i, queue, ioMap);
+  }
+
+  // Execute the algorithm in the order specified by queue.
+  const std::string algorithmName = getProperty(PropertyNames::ALGORITHM);
+  while (!queue.empty()) {
+    const auto row = queue.front();
+    auto &algorithmFactory = AlgorithmFactory::Instance();
```
Comment (DimitarTasev) [inline Framework/Algorithms/src/WorkflowAlgorithmRunner.cpp]: Can this be moved out of the `while` loop?

File: Framework/Algorithms/src/WorkflowAlgorithmRunner.cpp [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+void WorkflowAlgorithmRunner::configureRow(
+    ITableWorkspace_sptr setupTable, ITableWorkspace_sptr propertyTable,
+    const size_t currentRow, QUEUE &queue, const MAP &ioMap,
+    std::shared_ptr<std::unordered_set<size_t>> rowsBeingQueued) const {
+  if (currentRow > setupTable->rowCount()) {
+    throw std::runtime_error("Current row " + std::to_string(currentRow) +
+                             " out of task table bounds " +
+                             std::to_string(setupTable->rowCount()) + '.');
+  }
+  if (std::find(queue.cbegin(), queue.cend(), currentRow) != queue.cend()) {
```
Comment (DimitarTasev) [inline Framework/Algorithms/src/WorkflowAlgorithmRunner.cpp]: Can you put a comment here explaining what the intent of that if statement is, as I believe it's hard to get from a simple glance, and a comment would do nice?

**Review by dtasev — COMMENTED** [review]
- Works with different algorithm to example.
- Handles circular algorithms connections
- Documentation looks good
- Added to Release Notes
- Unit tests pass successfully

File: Framework/Algorithms/src/WorkflowAlgorithmRunner.cpp (hunk 1 above)
Comment (soininen) [inline Framework/Algorithms/src/WorkflowAlgorithmRunner.cpp]: Yes.

File: Framework/Algorithms/src/WorkflowAlgorithmRunner.cpp (hunk 2 above)
Comment (soininen) [inline Framework/Algorithms/src/WorkflowAlgorithmRunner.cpp]: Indeed. I added a bunch of comments to the method to hopefully make it more comprehensible.

**Review by dtasev — APPROVED** [review]
Everything looks good, :shipit: when builds pass

**Comment by SimonHeybrock** [comment]
Interesting!

**Pull request merged**